core/Services/Calendar/Store.vala

```diff
@@ -343,145 +694,408 @@ public class Calendar.Store : Object {
+            return;
+        }
+        debug ("Disconnecting source '%s'", source.dup_display_name ());
+        var views_removed = source_get_views (source);
+
+        var views_connected = source_get_views (source);
+        var views_connected_iterator = views_connected.iterator ();
+        while (views_connected_iterator.next ()) {
+            var view = views_connected_iterator.get ();
             try {
-                client.get_view.end (results, out view);
-                view.objects_added.connect ((objects) => on_objects_added (source, client, objects));
-                view.objects_removed.connect ((objects) => on_objects_removed (source, client, objects));
-                view.objects_modified.connect ((objects) => on_objects_modified (source, client, objects));
-                view.start ();
+                view_remove (view);
             } catch (Error e) {
-                critical ("Error from source '%s': %s", source.dup_display_name (), e.message);
+                error_received (e);
+                warning (e.message);
             }
+            views_connected = source_get_views (source);
+            views_connected_iterator = views_connected.iterator ();
+        }
 
-            source_view.set (source.dup_uid (), view);
-        });
+        lock (source_views) {
+            source_views.remove (source_uid);
+        }
+
+        lock (source_client) {
+            source_client.remove (source_uid);
+        }
+
+        var components = source_components.get (source_uid).get_values ().read_only_view;
+        components_removed (components, source, views_removed.read_only_view);
+        source_components.remove (source_uid);
+
+        source_removed (source);
     }
 
+    //--- Private Component EDS Event Handlers ---//
 
-    private async void add_source_async (E.Source source) {
-        debug ("Adding source '%s'", source.dup_display_name ());
-        try {
-            var cancellable = new GLib.Cancellable ();
-            connecting (source, cancellable);
-            var client = (ECal.Client) yield ECal.Client.connect (source, ECal.ClientSourceType.EVENTS, 30, cancellable);
-            source_client.insert (source.get_uid (), client);
-        } catch (Error e) {
-            error_received (e.message);
+    private async void source_component_add (E.Source source, ECal.Component component) throws Error {
+        unowned ICal.Component comp = component.get_icalcomponent ();
+        debug (@"Adding component '$(comp.get_uid())'");
+
+        ECal.Client? client;
+        lock (source_client) {
+            client = source_client.get (source.get_uid ());
         }
 
-        Idle.add (() => {
-            connected (source);
-            load_source (source);
-            return false;
-        });
+        if (client == null) {
+            critical ("No client was found for source '%s'", source.dup_display_name ());
+        } else {
+            string? uid;
```

Comment:
You don't do anything with `uid` here. The original code has a check to make sure the uid is set right after the `#endif`:
```vala
if (uid != null) {
    comp.set_uid (uid);
}
```

Reply:
Good catch! Although I'm not quite sure a) why this is needed and b) events are still fired as expected after this is re-added.

Re-added nevertheless for now, because it was there in the original code.